### ISSUE TITLE
Adding article on behalf of user via token access not possible

### ISSUE THREAD

**Issue by MrGeneration** [zammad/zammad#3327]
### Infos:

* Used Zammad version:
* Installation method (source, package, ..):
* Operating system:
* Database + version:
* Elasticsearch version:
* Browser + version:
* Ticket: #1081712

### Expected behavior:

Adding new articles on behalf of a customer works with either basic authentication or token access.

### Actual behavior:

Zammad returns "not authorized" if you try to add a new ticket article on behalf of a user with token access.

### Steps to reproduce the behavior:

* Try to add a new ticket article on behalf a customer using token auth

#### Used Payload:

```
curl --location --request POST 'https://zammad.tld/api/v1/ticket_articles' \
--header 'X-On-Behalf-Of: emily@example.com' \
--header 'Authorization: Bearer MySecretToken' \
--header 'Content-Type: application/json' \
--data-raw ' {
    "ticket_id": 9,
    "to": "",
    "cc": "",
    "subject": "some subject",
    "body": "test",
    "content_type": "text/html",
    "type": "web"
}'
```

#### Logged error:

```
I, [2020-12-07T15:05:35.812796 #14645-90075960]  INFO -- : Started POST "/api/v1/ticket_articles" for 666.666.666.666 at 2020-12-07 15:05:35 +0100
I, [2020-12-07T15:05:35.817442 #14645-90075960]  INFO -- : Processing by TicketArticlesController#create as */*
I, [2020-12-07T15:05:35.817506 #14645-90075960]  INFO -- :   Parameters: {"ticket_id"=>9, "to"=>"", "cc"=>"", "subject"=>"some subject", "body"=>"[FILTERED]", "content_type"=>"text/html", "type"=>"web"}
I, [2020-12-07T15:05:35.858264 #14645-90075960]  INFO -- : not allowed to create? this #<Ticket id: 9, group_id: 1, priority_id: 3, state_id: 2, organization_id: 4, number: "83009", title: "Need more information!", owner_id: 1, customer_id: 11, note: nil, first_response_at: nil, first_response_escalation_at: nil, first_response_in_min: nil, first_response_diff_in_min: nil, close_at: nil, close_escalation_at: nil, close_in_min: nil, close_diff_in_min: nil, update_escalation_at: nil, update_in_min: 0, update_diff_in_min: nil, last_contact_at: "2020-05-22 02:04:02", last_contact_agent_at: "2020-05-22 02:04:02", last_contact_customer_at: "2020-05-21 23:19:02", last_owner_update_at: nil, create_article_type_id: 1, create_article_sender_id: 2, article_count: 4, escalation_at: nil, pending_time: nil, time_unit: nil, preferences: {"escalation_calculation"=>{"escalation_disabled"=>false}}, updated_by_id: 4, created_by_id: 11, created_at: "2020-05-21 23:19:02", updated_at: "2020-11-19 08:25:32"> (Pundit::NotAuthorizedError)
/usr/local/rvm/gems/ruby-2.6.6/gems/pundit-2.0.1/lib/pundit.rb:221:in `authorize'
/opt/zammad/app/controllers/application_controller/authorizes.rb:8:in `authorize!'
/opt/zammad/app/controllers/ticket_articles_controller.rb:81:in `create'
/usr/local/rvm/gems/ruby-2.6.6/gems/actionpack-5.2.4.4/lib/action_controller/metal/basic_implicit_render.rb:6:in `send_action'
/usr/local/rvm/gems/ruby-2.6.6/gems/actionpack-5.2.4.4/lib/abstract_controller/base.rb:194:in `process_action'
/usr/local/rvm/gems/ruby-2.6.6/gems/actionpack-5.2.4.4/lib/action_controller/metal/rendering.rb:30:in `process_action'
/usr/local/rvm/gems/ruby-2.6.6/gems/actionpack-5.2.4.4/lib/abstract_controller/callbacks.rb:42:in `block in process_action'
/usr/local/rvm/gems/ruby-2.6.6/gems/activesupport-5.2.4.4/lib/active_support/callbacks.rb:109:in `block in run_callbacks'
/opt/zammad/app/controllers/application_controller/has_secure_content_security_policy_for_downloads.rb:18:in `block (4 levels) in <module:HasSecureContentSecurityPolicyForDownloads>'
/usr/local/rvm/gems/ruby-2.6.6/gems/activesupport-5.2.4.4/lib/active_support/notifications.rb:180:in `subscribed'
/opt/zammad/app/controllers/application_controller/has_secure_content_security_policy_for_downloads.rb:17:in `block (3 levels) in <module:HasSecureContentSecurityPolicyForDownloads>'
/usr/local/rvm/gems/ruby-2.6.6/gems/activesupport-5.2.4.4/lib/active_support/notifications.rb:180:in `subscribed'
/opt/zammad/app/controllers/application_controller/has_secure_content_security_policy_for_downloads.rb:16:in `block (2 levels) in <module:HasSecureContentSecurityPolicyForDownloads>'
/usr/local/rvm/gems/ruby-2.6.6/gems/activesupport-5.2.4.4/lib/active_support/callbacks.rb:118:in `instance_exec'
/usr/local/rvm/gems/ruby-2.6.6/gems/activesupport-5.2.4.4/lib/active_support/callbacks.rb:118:in `block in run_callbacks'
/usr/local/rvm/gems/ruby-2.6.6/gems/activesupport-5.2.4.4/lib/active_support/callbacks.rb:136:in `run_callbacks'
/usr/local/rvm/gems/ruby-2.6.6/gems/actionpack-5.2.4.4/lib/abstract_controller/callbacks.rb:41:in `process_action'
/usr/local/rvm/gems/ruby-2.6.6/gems/actionpack-5.2.4.4/lib/action_controller/metal/rescue.rb:22:in `process_action'
/usr/local/rvm/gems/ruby-2.6.6/gems/actionpack-5.2.4.4/lib/action_controller/metal/instrumentation.rb:34:in `block in process_action'
/usr/local/rvm/gems/ruby-2.6.6/gems/activesupport-5.2.4.4/lib/active_support/notifications.rb:168:in `block in instrument'
/usr/local/rvm/gems/ruby-2.6.6/gems/activesupport-5.2.4.4/lib/active_support/notifications/instrumenter.rb:23:in `instrument'
/usr/local/rvm/gems/ruby-2.6.6/gems/activesupport-5.2.4.4/lib/active_support/notifications.rb:168:in `instrument'
/usr/local/rvm/gems/ruby-2.6.6/gems/actionpack-5.2.4.4/lib/action_controller/metal/instrumentation.rb:32:in `process_action'
/usr/local/rvm/gems/ruby-2.6.6/gems/actionpack-5.2.4.4/lib/action_controller/metal/params_wrapper.rb:256:in `process_action'
/usr/local/rvm/gems/ruby-2.6.6/gems/activerecord-5.2.4.4/lib/active_record/railties/controller_runtime.rb:24:in `process_action'
/usr/local/rvm/gems/ruby-2.6.6/gems/actionpack-5.2.4.4/lib/abstract_controller/base.rb:134:in `process'
/usr/local/rvm/gems/ruby-2.6.6/gems/actionview-5.2.4.4/lib/action_view/rendering.rb:32:in `process'
/usr/local/rvm/gems/ruby-2.6.6/gems/actionpack-5.2.4.4/lib/action_controller/metal.rb:191:in `dispatch'
/usr/local/rvm/gems/ruby-2.6.6/gems/actionpack-5.2.4.4/lib/action_controller/metal.rb:252:in `dispatch'
/usr/local/rvm/gems/ruby-2.6.6/gems/actionpack-5.2.4.4/lib/action_dispatch/routing/route_set.rb:52:in `dispatch'
/usr/local/rvm/gems/ruby-2.6.6/gems/actionpack-5.2.4.4/lib/action_dispatch/routing/route_set.rb:34:in `serve'
/usr/local/rvm/gems/ruby-2.6.6/gems/actionpack-5.2.4.4/lib/action_dispatch/journey/router.rb:52:in `block in serve'
/usr/local/rvm/gems/ruby-2.6.6/gems/actionpack-5.2.4.4/lib/action_dispatch/journey/router.rb:35:in `each'
/usr/local/rvm/gems/ruby-2.6.6/gems/actionpack-5.2.4.4/lib/action_dispatch/journey/router.rb:35:in `serve'
/usr/local/rvm/gems/ruby-2.6.6/gems/actionpack-5.2.4.4/lib/action_dispatch/routing/route_set.rb:840:in `call'
/usr/local/rvm/gems/ruby-2.6.6/gems/omniauth-1.9.1/lib/omniauth/strategy.rb:420:in `call_app!'
/usr/local/rvm/gems/ruby-2.6.6/gems/omniauth-saml-1.10.1/lib/omniauth/strategies/saml.rb:89:in `other_phase'
/usr/local/rvm/gems/ruby-2.6.6/gems/omniauth-1.9.1/lib/omniauth/strategy.rb:190:in `call!'
/usr/local/rvm/gems/ruby-2.6.6/gems/omniauth-1.9.1/lib/omniauth/strategy.rb:169:in `call'
/usr/local/rvm/gems/ruby-2.6.6/gems/omniauth-1.9.1/lib/omniauth/strategy.rb:192:in `call!'
/usr/local/rvm/gems/ruby-2.6.6/gems/omniauth-1.9.1/lib/omniauth/strategy.rb:169:in `call'
/usr/local/rvm/gems/ruby-2.6.6/gems/omniauth-1.9.1/lib/omniauth/strategy.rb:192:in `call!'
/usr/local/rvm/gems/ruby-2.6.6/gems/omniauth-1.9.1/lib/omniauth/strategy.rb:169:in `call'
/usr/local/rvm/gems/ruby-2.6.6/gems/omniauth-1.9.1/lib/omniauth/strategy.rb:192:in `call!'
/usr/local/rvm/gems/ruby-2.6.6/gems/omniauth-1.9.1/lib/omniauth/strategy.rb:169:in `call'
/usr/local/rvm/gems/ruby-2.6.6/gems/omniauth-1.9.1/lib/omniauth/strategy.rb:192:in `call!'
/usr/local/rvm/gems/ruby-2.6.6/gems/omniauth-1.9.1/lib/omniauth/strategy.rb:169:in `call'
/usr/local/rvm/gems/ruby-2.6.6/gems/omniauth-1.9.1/lib/omniauth/strategy.rb:192:in `call!'
/usr/local/rvm/gems/ruby-2.6.6/gems/omniauth-1.9.1/lib/omniauth/strategy.rb:169:in `call'
/usr/local/rvm/gems/ruby-2.6.6/gems/omniauth-1.9.1/lib/omniauth/strategy.rb:192:in `call!'
/usr/local/rvm/gems/ruby-2.6.6/gems/omniauth-1.9.1/lib/omniauth/strategy.rb:169:in `call'
/usr/local/rvm/gems/ruby-2.6.6/gems/omniauth-1.9.1/lib/omniauth/strategy.rb:192:in `call!'
/usr/local/rvm/gems/ruby-2.6.6/gems/omniauth-1.9.1/lib/omniauth/strategy.rb:169:in `call'
/usr/local/rvm/gems/ruby-2.6.6/gems/omniauth-1.9.1/lib/omniauth/strategy.rb:192:in `call!'
/usr/local/rvm/gems/ruby-2.6.6/gems/omniauth-1.9.1/lib/omniauth/strategy.rb:169:in `call'
/usr/local/rvm/gems/ruby-2.6.6/gems/omniauth-1.9.1/lib/omniauth/strategy.rb:192:in `call!'
/usr/local/rvm/gems/ruby-2.6.6/gems/omniauth-1.9.1/lib/omniauth/strategy.rb:169:in `call'
/usr/local/rvm/gems/ruby-2.6.6/gems/omniauth-1.9.1/lib/omniauth/builder.rb:45:in `call'
/usr/local/rvm/gems/ruby-2.6.6/gems/rack-2.2.3/lib/rack/tempfile_reaper.rb:15:in `call'
/usr/local/rvm/gems/ruby-2.6.6/gems/rack-2.2.3/lib/rack/etag.rb:27:in `call'
/usr/local/rvm/gems/ruby-2.6.6/gems/rack-2.2.3/lib/rack/conditional_get.rb:40:in `call'
/usr/local/rvm/gems/ruby-2.6.6/gems/rack-2.2.3/lib/rack/head.rb:12:in `call'
/usr/local/rvm/gems/ruby-2.6.6/gems/actionpack-5.2.4.4/lib/action_dispatch/http/content_security_policy.rb:18:in `call'
/usr/local/rvm/gems/ruby-2.6.6/gems/rack-2.2.3/lib/rack/session/abstract/id.rb:266:in `context'
/usr/local/rvm/gems/ruby-2.6.6/gems/rack-2.2.3/lib/rack/session/abstract/id.rb:260:in `call'
/usr/local/rvm/gems/ruby-2.6.6/gems/actionpack-5.2.4.4/lib/action_dispatch/middleware/cookies.rb:670:in `call'
/usr/local/rvm/gems/ruby-2.6.6/gems/actionpack-5.2.4.4/lib/action_dispatch/middleware/callbacks.rb:28:in `block in call'
/usr/local/rvm/gems/ruby-2.6.6/gems/activesupport-5.2.4.4/lib/active_support/callbacks.rb:98:in `run_callbacks'
/usr/local/rvm/gems/ruby-2.6.6/gems/actionpack-5.2.4.4/lib/action_dispatch/middleware/callbacks.rb:26:in `call'
/usr/local/rvm/gems/ruby-2.6.6/gems/actionpack-5.2.4.4/lib/action_dispatch/middleware/debug_exceptions.rb:61:in `call'
/usr/local/rvm/gems/ruby-2.6.6/gems/actionpack-5.2.4.4/lib/action_dispatch/middleware/show_exceptions.rb:33:in `call'
/usr/local/rvm/gems/ruby-2.6.6/gems/railties-5.2.4.4/lib/rails/rack/logger.rb:38:in `call_app'
/usr/local/rvm/gems/ruby-2.6.6/gems/railties-5.2.4.4/lib/rails/rack/logger.rb:26:in `block in call'
/usr/local/rvm/gems/ruby-2.6.6/gems/activesupport-5.2.4.4/lib/active_support/tagged_logging.rb:71:in `block in tagged'
/usr/local/rvm/gems/ruby-2.6.6/gems/activesupport-5.2.4.4/lib/active_support/tagged_logging.rb:28:in `tagged'
/usr/local/rvm/gems/ruby-2.6.6/gems/activesupport-5.2.4.4/lib/active_support/tagged_logging.rb:71:in `tagged'
/usr/local/rvm/gems/ruby-2.6.6/gems/railties-5.2.4.4/lib/rails/rack/logger.rb:26:in `call'
/usr/local/rvm/gems/ruby-2.6.6/gems/actionpack-5.2.4.4/lib/action_dispatch/middleware/remote_ip.rb:81:in `call'
/usr/local/rvm/gems/ruby-2.6.6/gems/actionpack-5.2.4.4/lib/action_dispatch/middleware/request_id.rb:27:in `call'
/usr/local/rvm/gems/ruby-2.6.6/gems/rack-2.2.3/lib/rack/method_override.rb:24:in `call'
/usr/local/rvm/gems/ruby-2.6.6/gems/rack-2.2.3/lib/rack/runtime.rb:22:in `call'
/usr/local/rvm/gems/ruby-2.6.6/gems/activesupport-5.2.4.4/lib/active_support/cache/strategy/local_cache_middleware.rb:29:in `call'
/usr/local/rvm/gems/ruby-2.6.6/gems/actionpack-5.2.4.4/lib/action_dispatch/middleware/executor.rb:14:in `call'
/usr/local/rvm/gems/ruby-2.6.6/gems/actionpack-5.2.4.4/lib/action_dispatch/middleware/static.rb:127:in `call'
/usr/local/rvm/gems/ruby-2.6.6/gems/rack-2.2.3/lib/rack/sendfile.rb:110:in `call'
/usr/local/rvm/gems/ruby-2.6.6/gems/railties-5.2.4.4/lib/rails/engine.rb:524:in `call'
/usr/local/rvm/gems/ruby-2.6.6/gems/railties-5.2.4.4/lib/rails/railtie.rb:190:in `public_send'
/usr/local/rvm/gems/ruby-2.6.6/gems/railties-5.2.4.4/lib/rails/railtie.rb:190:in `method_missing'
/usr/local/rvm/gems/ruby-2.6.6/gems/puma-3.12.6/lib/puma/configuration.rb:227:in `call'
/usr/local/rvm/gems/ruby-2.6.6/gems/puma-3.12.6/lib/puma/server.rb:706:in `handle_request'
/usr/local/rvm/gems/ruby-2.6.6/gems/puma-3.12.6/lib/puma/server.rb:476:in `process_client'
/usr/local/rvm/gems/ruby-2.6.6/gems/puma-3.12.6/lib/puma/server.rb:334:in `block in run'
/usr/local/rvm/gems/ruby-2.6.6/gems/puma-3.12.6/lib/puma/thread_pool.rb:135:in `block in spawn_thread'
/usr/local/rvm/gems/ruby-2.6.6/gems/logging-2.2.2/lib/logging/diagnostic_context.rb:474:in `block in create_with_logging_context'
I, [2020-12-07T15:05:35.858453 #14645-90075960]  INFO -- : Exceptions::NotAuthorized (Exceptions::NotAuthorized)
I, [2020-12-07T15:05:35.858809 #14645-90075960]  INFO -- : Completed 401 Unauthorized in 41ms (Views: 0.2ms | ActiveRecord: 24.8ms)
```

Yes I'm sure this is a bug and no feature request or a general question.

**Comment by rkreienbuehl** [comment]
Thanks for adding this bug.
As mentioned in the documentation `This functionality can be used for any type of action.`, the on behalf header should work on every action, but also on other requests, like list the articles or show a specific article, I get a `401 Unauthorized` back, not only on creating new article.

**Comment by rkreienbuehl** [comment]
Is there any development update on this bug?

**Comment by MrGeneration** [comment]
Please don't push issues like that.
We will update this issue accordingly as soon as anything moves.

Please note that we can't and won't provide ETAs in general.

**Comment by thorsteneckel** [comment]
We need more information on this:
- What are the permissions of the `Bearer Token` user that is performing the request?
- What are the permissions of the `X-On-Behalf-Of` user? (`ticket.agent`/`ticket.customer`)
- What are the permissions configured for the Token?

If the `X-On-Behalf-Of` user has `ticket.customer` permissions:
- Has the `X-On-Behalf-Of` user customer permissions to the ticket? (Customer or shared Organization access)?

**Comment by rkreienbuehl** [comment]
> * What are the permissions of the `Bearer Token` user that is performing the request?

The Token is not a `Bearer Token`, its a personal Access Token and we use it in the Header with `Authorization: Token token={your_token}`
The permissions for the Token are:
`admin.ticket`, `admin.user` and `ticket.agent`

> * What are the permissions of the `X-On-Behalf-Of` user? (`ticket.agent`/`ticket.customer`)

The `X-On-Behalf-Of` is a normal User (that was create over the API if he didn't existed) and has the rights of a customer.

> * What are the permissions configured for the Token?

`admin.ticket`, `admin.user` and `ticket.agent`

> * Has the `X-On-Behalf-Of` user customer permissions to the ticket? (Customer or shared Organization access)?

Yes, we only provide access to tickets where this user is involved. Mostly, the ticket will be generated over the API by this user.

**Comment by thorsteneckel** [comment]
Thanks a lot @rkreienbuehl - with that information I was able to reproduce the issue 💪 I translated it into the following RSpec test code:

```ruby
token_owner = create(:admin, firstname: 'Owner')
token       = create(:token, user: token_owner, preferences: { permission: %w[admin.ticket admin.user ticket.agent] })
behalf_of   = create(:customer, firstname: 'Behalf of')

ticket = create(:ticket, customer: behalf_of)

authenticated_as(token_owner, via: :api_client, token: token, on_behalf_of: behalf_of.email)

params = {
  ticket_id:    ticket.id,
  to:           '',
  cc:           '',
  subject:      'some subject',
  body:         'test',
  content_type: 'text/html',
  type:         'web',
}
post '/api/v1/ticket_articles', params: params, as: :json
```
Side note: The `authenticated_as` helper needed an adjustment which I will commit later.

After some debugging I found out that this works as intended. The Token is used in the `ticket.customer` permission context - as it needs to create an Article in the name of / with the access permissions of the Customer (`X-On-Behalf-Of` user). However, neither the Token Admin/Agent-User nor the Token itself have those permissions.

The check is performed here:
https://github.com/zammad/zammad/blob/2ad41b639e5100abed1d9bb9cd259fb6d4712342/app/models/token.rb#L107-L111

Where `effective_user.permissions?('ticket.customer')` (`effective_user` == `X-On-Behalf-Of` User) is `true` but `super('ticket.customer')` (read as: `token.permissions?('ticket.customer')`) is `false` and therefore fails the authorization.

This behaviour was introduced by fixing #3192 and #3186.

To achieve the desired behaviour assign `ticket.customer` permissions to both the Requestion Admin/Agent User and the Token.

Closing now.
